integration-tests/src/test/java/com/linkedin/kafka/clients/producer/LiKafkaProducerIntegrationTest.java

```diff
@@ -195,5 +209,79 @@ public void close() {
     assertEquals(1, messageCount);
   }
 
+  @Test
+  public void testCustomPartitioningOfLargeMessage() throws Exception {
+    long seed = System.currentTimeMillis();
+    Random random = new Random(seed);
+    int numPartitions = 2 + random.nextInt(19); // [2-20]
+    String topic = "testCustomPartitioner-" + UUID.randomUUID();
+
+    AdminClient adminClient = createRawAdminClient(null);
+    CreateTopicsResult createTopicResult = adminClient.createTopics(Collections.singletonList(new NewTopic(topic, numPartitions, (short) 1)));
+    createTopicResult.all().get(1, TimeUnit.MINUTES);
+    TopicDescription topicDescription = adminClient.describeTopics(Collections.singletonList(topic)).all().get(1, TimeUnit.MINUTES).get(topic);
+    Assert.assertEquals(topicDescription.partitions().size(), numPartitions, "expected topic to have "
+        + numPartitions + " partitions. instead got " + topicDescription.partitions() + ". seed is " + seed);
+
+    Properties producerPropertyOverrides = new Properties();
+    producerPropertyOverrides.put(ProducerConfig.PARTITIONER_CLASS_CONFIG, RoundRobinPartitioner.class.getCanonicalName());
+    producerPropertyOverrides.put(LiKafkaProducerConfig.LARGE_MESSAGE_ENABLED_CONFIG, "true");
+
+    int msgSize = (int) (1.5 * DEFAULT_MAX_SEGMENT_BYTES);
+    int numExpectedRecords = (int) Math.ceil(((double) msgSize) / DEFAULT_MAX_SEGMENT_BYTES);
+    String randomLargeString = RandomStringUtils.randomAlphanumeric(msgSize);
+
+    try (LiKafkaProducer<String, String> producer =  createProducer(producerPropertyOverrides)) {
+      //no explicit partition set, timestamp = now
+      ProducerRecord<String, String> record = new ProducerRecord<>(topic, null, System.currentTimeMillis(), "key", randomLargeString);
+      producer.send(record).get();
+    }
+
+    List<TopicPartition> tps = topicDescription.partitions().stream()
+        .map(tpi -> new TopicPartition(topic, tpi.partition()))
+        .collect(Collectors.toList());
+
+    try (Consumer<byte[], byte[]> consumer = createRawConsumer()) {
+      Map<Integer, Integer> partToCount = new HashMap<>(); //histogram of partition to how many records seen on it
+      int recordsRead = 0;
+      consumer.assign(tps);
+      consumer.seekToBeginning(tps);
+      long deadline = System.currentTimeMillis() + TimeUnit.MINUTES.toMillis(2);
+      while (System.currentTimeMillis() < deadline && recordsRead < numExpectedRecords) {
+        ConsumerRecords<byte[], byte[]> records = consumer.poll(Duration.ofSeconds(1));
+        recordsRead += records.count();
+        for (ConsumerRecord<byte[], byte[]> rec : records) {
+          partToCount.compute(rec.partition(), (k, v) -> {
+            if (v == null) {
+              return 1;
+            }
+            return v + 1;
+          });
+        }
+      }
+      Assert.assertEquals(recordsRead, numExpectedRecords, "only read " + recordsRead + " within timeout. seed is " + seed);
```

Comment:
This assertion can be merged with line#265. 

Assert.assertEquals(partToCount.values.iterator.next(), numExpectedRecrods, "...")